### ISSUE TITLE
Reuse layers at prediction time

### ISSUE THREAD

**Issue by Guillermogsjc** [IndicoDataSolutions/finetune#325]
Hi,

there is a feature that would be nice to have (I do not know exactly if you already integrated it).

Lets suppose that we train several models for different tasks on the same data. Every model just training the 3 last layers for the task.

At prediction time, instead of performing  full `.predict` with all the models, it would be very nice to factorize the input pipeline and the 9 first layer transformations, getting this 9th layers, then applying only 3 layers of each particular model trained (and already on memory).

It would save a lot of time when serving the model. Probably just adding this factorization to the example https://finetune.indico.io/#using-adapters-and-the-deploymentmodel-class. The idea is to have every "adaptation" of the model on memory, perform the  fordward transformations on the data until the adapters start, then apply every "adaptation" already loaded over this point.

Regards.

**Comment by matthewbayer** [comment]
Hi Guillermogsjc -

Thanks for suggesting this feature - we agree that this factorization is doable given our current DeploymentModel structure. However, we do not plan to support this due to the fact that our own experiments, in addition to those of the paper Parameter Efficient Finetuning for Efficient NLP (where adapters were proposed), show that adapters provide better performance than models that retrain only the last few layers. Because of this, we want encourage use of adapters with the DeploymentModel because they have lower parameter counts but retain model accuracy, which best leverages the features of the DeploymentModel.

However, if you would like to add this solution to Finetune, we are happy to get in contact help you out along the way.

Best of luck!

**Comment by Guillermogsjc** [comment]
Thanks Matthew, will explore better the adapters.

Regards.
